File: tests/startup-config.test.ts
import { describe, expect, it } from "vitest";

import { normalizeBaseUrl, parseStartupConfig } from "../src/startup-config.js";

describe("parseStartupConfig", () => {
  it("reads the base URL from CLI arguments", () => {
    expect(parseStartupConfig(["--base-url", "https://docs.example.com/"])).toEqual({
      baseUrl: "https://docs.example.com",
    });
  });

  it("supports inline CLI argument syntax", () => {
    expect(parseStartupConfig(["--base-url=https://docs.example.com/"])).toEqual({
      baseUrl: "https://docs.example.com",
    });
  });

  it("falls back to the environment when no CLI argument is provided", () => {
    expect(parseStartupConfig([], { DOCMOST_BASE_URL: "https://env.example.com/" })).toEqual({
      baseUrl: "https://env.example.com",
    });
  });

  it("throws when the base URL flag is missing a value", () => {
    expect(() => parseStartupConfig(["--base-url"])).toThrow("Missing value for --base-url.");
  });
});

describe("normalizeBaseUrl", () => {
  it("removes trailing slashes", () => {
    expect(normalizeBaseUrl("https://docs.example.com///")).toBe("https://docs.example.com");
  });
});

File: tests/startup_config_test.rs
use std::collections::HashMap;

use docmost_local_mcp::startup_config::{normalize_base_url, parse_startup_config};

#[test]
fn reads_base_url_from_cli_arguments() {
    let argv = vec![
        "--base-url".to_string(),
        "https://docs.example.com/".to_string(),
    ];
    let config = parse_startup_config(&argv, &HashMap::new()).unwrap();
    assert_eq!(config.base_url.as_deref(), Some("https://docs.example.com"));
}

#[test]
fn supports_inline_cli_argument_syntax() {
    let argv = vec!["--base-url=https://docs.example.com/".to_string()];
    let config = parse_startup_config(&argv, &HashMap::new()).unwrap();
    assert_eq!(config.base_url.as_deref(), Some("https://docs.example.com"));
}

#[test]
fn falls_back_to_environment() {
    let env = HashMap::from([(
        "DOCMOST_BASE_URL".to_string(),
        "https://env.example.com/".to_string(),
    )]);
    let config = parse_startup_config(&[], &env).unwrap();
    assert_eq!(config.base_url.as_deref(), Some("https://env.example.com"));
}

#[test]
fn throws_when_base_url_flag_is_missing_value() {
    let argv = vec!["--base-url".to_string()];
    let error = parse_startup_config(&argv, &HashMap::new()).unwrap_err();
    assert_eq!(error.to_string(), "Missing value for --base-url.");
}

#[test]
fn removes_trailing_slashes() {
    assert_eq!(
        normalize_base_url("https://docs.example.com///"),
        "https://docs.example.com"
    );
}
